ingest/usecase/plugins/orderbookfiller/ordebook_filler_ingest_plugin.go

```diff
@@ -0,0 +1,180 @@
+package orderbookfiller
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+	"github.com/osmosis-labs/sqs/domain"
+	"github.com/osmosis-labs/sqs/domain/keyring"
+	"github.com/osmosis-labs/sqs/domain/mvc"
+	"github.com/osmosis-labs/sqs/log"
+	"go.uber.org/zap"
+)
+
+// orderbookFillerIngestPlugin is a plugin that fills the orderbook orders at the end of the block.
+type orderbookFillerIngestPlugin struct {
+	poolsUseCase  mvc.PoolsUsecase
+	routerUseCase mvc.RouterUsecase
+	tokensUseCase mvc.TokensUsecase
+
+	atomicBool atomic.Bool
+
+	keyring           keyring.Keyring
+	defaultQuoteDenom string
+
+	logger log.Logger
+
+	swapDone atomic.Bool
+}
+
+var _ domain.EndBlockProcessPlugin = &orderbookFillerIngestPlugin{}
+
+func New(poolsUseCase mvc.PoolsUsecase, routerUseCase mvc.RouterUsecase, tokensUseCase mvc.TokensUsecase, keyring keyring.Keyring, defaultQuoteDenom string, logger log.Logger) *orderbookFillerIngestPlugin {
+	return &orderbookFillerIngestPlugin{
+		poolsUseCase:  poolsUseCase,
+		routerUseCase: routerUseCase,
+		tokensUseCase: tokensUseCase,
+
+		atomicBool: atomic.Bool{},
+
+		keyring:           keyring,
+		defaultQuoteDenom: defaultQuoteDenom,
+
+		logger: logger,
+	}
+}
+
+// ProcessEndBlock implements domain.EndBlockProcessPlugin.
+func (o *orderbookFillerIngestPlugin) ProcessEndBlock(ctx context.Context, blockHeight uint64, metadata domain.BlockPoolMetadata) error {
+	canonicalOrderbooks, err := o.poolsUseCase.GetAllCanonicalOrderbookPoolIDs()
+	if err != nil {
+		o.logger.Error("failed to get all canonical orderbook pool IDs", zap.Error(err))
+		return err
+	}
+
+	resultChan := make(chan error, len(canonicalOrderbooks))
+	defer close(resultChan)
+
+	for _, canonicalOrderbook := range canonicalOrderbooks {
+		go func(canonicalOrderbook domain.CanonicalOrderBooksResult) {
+			resultChan <- o.processOrderbook(ctx, canonicalOrderbook)
+		}(canonicalOrderbook)
+	}
+
+	// Collect all the results
+	for i := 0; i < len(canonicalOrderbooks); i++ {
+		select {
+		case err := <-resultChan:
+			if err != nil {
+				o.logger.Debug("failed to process orderbook", zap.Error(err))
+			}
+		case <-time.After(10 * time.Second):
+			o.logger.Error("timed out processing orderbook")
+			return err
+		}
+	}
+
+	o.logger.Info("processed end block in orderbook filler ingest plugin", zap.Uint64("block_height", blockHeight))
+	return nil
```

Comment:
@niccoloraspa we read the canonical orderbooks and then try to detect cyclic arb opportunities against top SQS route for each